rotkehlchen/tasks/assets.py

```diff
@@ -381,3 +390,51 @@ def update_aave_v3_underlying_assets(chains_aggregator: 'ChainsAggregator') -> N
             name=DBCacheStatic.LAST_AAVE_V3_ASSETS_UPDATE,
             value=ts_now(),
         )
+
+
+def maybe_detect_new_tokens(database: 'DBHandler') -> None:
+    if not CachedSettings().get_settings().auto_detect_tokens:
+        return
+
+    with database.conn.write_ctx() as write_cursor:
+        last_save_time = database.get_last_balance_save_time(write_cursor)
+        detected_tokens = defaultdict(list)
+        # we query the decoded history events that are the first event of each distinct
+        # unignored asset for that account that happened on or after last_save_time.
+        for event_data in write_cursor.execute(
+            # events that are the earliest events of distinct assets after last_save_time
+            f'SELECT {HISTORY_BASE_ENTRY_FIELDS}, '
+            f'{EVM_EVENT_FIELDS} {EVM_EVENT_JOIN} {filter_ignore_asset_query()} '
+            'AND timestamp >= ? GROUP BY asset, location_label;',
+            (ts_sec_to_ms(last_save_time),),
```

Comment:
you have detected tokens for only one chain at this point right? Wouldn't it be better to filter by chain? With that you could avoid the check `event.location not in EVM_LOCATIONS`

Reply:
Don't do it because you can loose events. The function taht you call at the end updates `last_save_time` and you can miss tokens... but then question: this logic will fail because 

1. YOu query gnosis and you have events between timstamps [17, 20]. You call this logic and `last_save_time = 22`
2. You query ethereum and you ahve events between timesatmp [18,21]. YOu call this logic and you do nothing since there are no events after `last_save_time = 22`

Is this expected? Not sure what is the objetive that you and left have regarding this

Reply:
now we call it outside of the loop from a different cursor, so should be fine now